src/named_pipe.rs

```diff
@@ -0,0 +1,208 @@
+use std::fs::File;
+use std::fs::OpenOptions;
+use std::os::raw::c_int;
+use std::os::unix::fs::OpenOptionsExt;
+use std::os::unix::io::AsRawFd;
+use std::os::unix::io::RawFd;
+use std::path::PathBuf;
+use std::sync::atomic::{AtomicU32, Ordering};
+use std::thread::sleep;
+use std::time::{Duration, Instant};
+
+use nix::errno::Errno;
+use nix::fcntl::OFlag;
+use nix::poll::{poll, PollFd, PollFlags};
+use nix::sys::stat::Mode;
+use nix::unistd::{mkfifo, read, write};
+use tempfile::TempDir;
+
+use crate::error::{WorkerError, WorkerErrorKind};
+
+#[derive(Debug)]
+pub struct NamedPipeCreator {
+    dir: TempDir,
+    counter: AtomicU32,
+}
+
+impl NamedPipeCreator {
+    pub fn new() -> Result<Self, WorkerError> {
+        Ok(Self {
+            dir: TempDir::new()?,
+            counter: AtomicU32::new(0),
+        })
+    }
+
+    pub fn new_pipe(&self) -> Result<NamedPipe, WorkerError> {
+        let pipe_num = self.counter.fetch_add(1, Ordering::SeqCst);
+
+        let component_input_fifo_filename = format!("IN_{}", pipe_num);
+        let component_input_fifo_path = self.dir.path().join(component_input_fifo_filename);
+
+        let component_output_fifo_filename = format!("OUT_{}", pipe_num);
+        let component_output_fifo_path = self.dir.path().join(component_output_fifo_filename);
+
+        mkfifo(
+            &component_input_fifo_path,
+            Mode::S_IRWXU | Mode::S_IRWXG | Mode::S_IRWXO,
+        )?;
+        mkfifo(
+            &component_output_fifo_path,
+            Mode::S_IRWXU | Mode::S_IRWXG | Mode::S_IRWXO,
+        )?;
+
+        debug!(
+            "Creating new pipe I = {:?}, O = {:?}",
+            component_input_fifo_path, component_output_fifo_path
+        );
+
+        Ok(NamedPipe {
+            component_input_fifo_path,
+            component_output_fifo_path,
+
+            component_input_fifo_file: None,
+            component_output_fifo_file: None,
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct NamedPipe {
+    component_input_fifo_path: PathBuf,
+    component_output_fifo_path: PathBuf,
+
+    component_input_fifo_file: Option<File>,
+    component_output_fifo_file: Option<File>,
+}
+
+const PIPE_CREATION_TIMEOUT_MS: u64 = 100;
+const PIPE_IO_TIMEOUT_MS: u64 = 1000;
+const PIPE_POLL_INTERVAL_MS: u64 = 2;
+
+const BUF_SIZE: usize = 256;
+
+impl NamedPipe {
+    fn get_fds(&mut self) -> Result<(RawFd, RawFd), WorkerError> {
+        let deadline = Instant::now() + Duration::from_millis(PIPE_CREATION_TIMEOUT_MS);
```

Comment:
Do we have this deadline because pipe creation might fail?

Reply:
No, it's because pipes won't let you open them until the other side is ready. We need to spin a bit until the process is ready